### PR TITLE
[JENKINS-41050] Always do SCM checkout directly on agent, not in container

### PR DESCRIPTION
**Pull request by abayer**
[JENKINS-41050](https://issues.jenkins-ci.org/browse/JENKINS-41050)

cc @reviewbybees esp @michaelneale @rsandell @HRMPW

### CODE REVIEW

**Comment by rsandell** [comment]
I put a comment in the ticket since I'm not sure this is the right thing to do.

**Comment by abayer** [comment]
@rsandell - FYI, I made some nontrivial changes here - some bungled changes from earlier that needed to be cleaned up post-merge of other stuff, and getting rid of the `declarativeProps` (previously `dockerLabel`) step entirely and just using whitelisted static methods. Much simpler, less noise (both in the logs - i.e., no more `[Pipeline] declarativeProps` - and in the flow graph, since this behavior really didn't merit a full blown `Step` and therefore its own `FlowNode` etc).

**Comment by michaelneale** [comment]
functionality wise :bee: from me - but not for code review.

**Comment by rsandell** [comment]
@abayer then I'd have to give you a :bug: those types of things don't belong in this change and needs to be done separately.

**Comment by abayer** [comment]
@rsandell fair enough. =)

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/agent/impl/AnyScript.groovy [hunk 1]
```diff
@@ -37,15 +36,11 @@ public class AnyScript extends DeclarativeAgentScript<Any> {
...
-        return {
-            try {
-                script.node {
-                    body.call()
-                }
-            } catch (Exception e) {
-                script.getProperty("currentBuild").result = Result.FAILURE
-                throw e
-            }
+        Label l = (Label) Label.DescriptorImpl.instanceForName("label", [label: null])
```
Comment (rsandell) [inline pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/agent/impl/AnyScript.groovy]: It looks overly abusive having to do `setContext` over and over again for the caller, can't that be a part of `instanceForName` ?

Comment (abayer) [inline pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/agent/impl/AnyScript.groovy]: Hmm...

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/agent/impl/LabelScript.groovy [hunk 2]
```diff
@@ -41,6 +45,16 @@ public class LabelScript extends DeclarativeAgentScript<Label> {
         return {
             try {
                 script.node(describable?.label) {
+                    if (describable.context instanceof Root) {
+                        Root root = (Root)describable.context
+                        SkipDefaultCheckout skip = (SkipDefaultCheckout)root.options?.options?.get("skipDefaultCheckout")
+                        if (!skip?.isSkipDefaultCheckout() && Utils.hasScmContext(script)) {
```
Comment (rsandell) [inline pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/agent/impl/LabelScript.groovy]: seems wrong to having this in here, but I understand the reasoning.

File: pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/steps/DeclarativePropsStepTest.java [hunk 3]
```diff
@@ -173,6 +173,7 @@ public void runsOnCorrectSlave() throws Exception {
         s.setLabelString("thisone");
         env(s).put("DOCKER_INDICATOR", "CORRECT").set();
         GlobalConfig.get().setDockerLabel("thisone");
+        GlobalConfig.get().setRegistry(null);
```
Comment (rsandell) [inline pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/steps/DeclarativePropsStepTest.java]: What does this have to do with it?

Comment (abayer) [inline pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/steps/DeclarativePropsStepTest.java]: Whoops, that should be over in #96...

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/agent/impl/AnyScript.groovy (hunk 1 above)
Comment (abayer) [inline pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/agent/impl/AnyScript.groovy]: Ah, duh. Fixing.

File: pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/steps/DeclarativePropsStepTest.java (hunk 3 above)
Comment (abayer) [inline pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/steps/DeclarativePropsStepTest.java]: Check that, it did need to be here. In fairness, I'm not entirely sure *why* it started failing here and not earlier, but...yeah.

Comment (abayer) [inline pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/steps/DeclarativePropsStepTest.java]: Argh, I read the wrong test results. Sigh.
